File: packages/pluto/src/function.ts
import {
  IResource,
  IResourceCapturedProps,
  IResourceInfraApi,
  IResourceClientApi,
  PlatformType,
  utils,
  FnResource,
  simulator,
} from "@plutolang/base";
import { aws, k8s, ali } from "./clients";

export type AnyFunction = (...args: any[]) => any;
export const DEFAULT_FUNCTION_NAME = "default";

/**
 * The interal protocol for the direct call response.
 */
export interface DirectCallResponse {
  // The status code of the response, same as the HTTP status code.
  code: number;
  // The result of the function call, or the error message.
  body: any;
}

interface FunctionHandler extends AnyFunction, FnResource {}

/**
 * The options for instantiating an infrastructure implementation class or a client implementation
 * class.
 */
export interface FunctionOptions {
  name?: string;
  envs?: Record<string, any>;
}

export interface IFunctionClientApi<T extends AnyFunction> extends IResourceClientApi {
  invoke(...args: Parameters<T>): Promise<Awaited<ReturnType<T> | void>>;
}

export interface IFunctionInfraApi extends IResourceInfraApi {}

export interface IFunctionCapturedProps extends IResourceCapturedProps {}

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * client implementation class of a resource type.
 */
export type IFunctionClient<T extends AnyFunction> = IFunctionClientApi<T> & IFunctionCapturedProps;

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * infrastructure implementation class of a resource type.
 */
export type IFunctionInfra = IFunctionInfraApi & IFunctionCapturedProps;

export class Function<T extends FunctionHandler> implements IResource {
  constructor(func: T, opts?: FunctionOptions) {
    func;
    opts;
    throw new Error(
      "Cannot instantiate this class, instead of its subclass depending on the target runtime."
    );
  }

  public static buildClient<T extends FunctionHandler>(
    func: T,
    opts?: FunctionOptions
  ): IFunctionClient<T> {
    const platformType = utils.currentPlatformType();
    switch (platformType) {
      case PlatformType.AWS:
        return new aws.LambdaFunction(func, opts);
      case PlatformType.K8s:
        return new k8s.KnativeService(func, opts);
      case PlatformType.AliCloud:
        return new ali.FCInstance(func, opts);
      case PlatformType.Simulator: {
        if (!process.env.PLUTO_SIMULATOR_URL) throw new Error("PLUTO_SIMULATOR_URL doesn't exist");
        const resourceId = utils.genResourceId(Function.fqn, opts?.name ?? DEFAULT_FUNCTION_NAME);
        return simulator.makeSimulatorClient(process.env.PLUTO_SIMULATOR_URL!, resourceId);
      }
      default:
        throw new Error(`not support this runtime '${platformType}'`);
    }
  }

  public static fqn = "@plutolang/pluto.Function";
}

export interface Function<T extends FunctionHandler>
  extends IResource,
    IFunctionClient<T>,
    IFunctionInfra {}

File: packages/pluto-py/pluto_client/function.py
from dataclasses import dataclass
from typing import Any, Callable, Dict, Generic, Optional, TypeVar
from pluto_base.resource import (
    IResource,
    IResourceCapturedProps,
    IResourceInfraApi,
    IResourceClientApi,
)
from pluto_base.platform import PlatformType
from pluto_base import utils


DEFAULT_FUNCTION_NAME = "default"

FnHandler = TypeVar("FnHandler", bound=Callable[..., Any])


@dataclass
class DirectCallResponse:
    code: int
    body: Any


@dataclass
class FunctionOptions:
    memory: int | None = 128  # The memory size in MB, default is 128.
    envs: Dict[str, Any] | None = None
    raw: bool = False  # This option only works for the AWS currently.


class IFunctionClientApi(Generic[FnHandler], IResourceClientApi):
    def invoke(self, *args, **kwargs) -> Any:
        raise NotImplementedError


class IFunctionInfraApi(IResourceInfraApi):
    pass


class IFunctionCapturedProps(IResourceCapturedProps):
    def url(self) -> str:
        raise NotImplementedError


class IFunctionClient(IFunctionClientApi[FnHandler], IFunctionCapturedProps):
    pass


class IFunctionInfra(IFunctionInfraApi, IFunctionCapturedProps):
    pass


class Function(IResource, IFunctionClient[FnHandler], IFunctionInfra):
    fqn = "@plutolang/pluto.Function"

    def __init__(
        self,
        func: FnHandler,
        name: Optional[str] = None,
        opts: Optional[FunctionOptions] = None,
    ):
        raise NotImplementedError(
            "Cannot instantiate this class, instead of its subclass depending on the target runtime."
        )

    @staticmethod
    def build_client(
        func: FnHandler,
        name: Optional[str] = None,
        opts: Optional[FunctionOptions] = None,
    ) -> IFunctionClient[FnHandler]:
        platform_type = utils.current_platform_type()
        if platform_type == PlatformType.AWS:
            from .clients import aws

            return aws.LambdaFunction(func, name, opts)
        else:
            raise ValueError(f"not support this runtime '{platform_type}'")
